iceoryx_posh/source/runtime/posh_runtime.cpp

```diff
@@ -484,6 +484,7 @@ cxx::expected<Error> PoshRuntime::findService(const capro::ServiceDescription& s
         return cxx::error<Error>(Error::kMQ_INTERFACE__REG_UNABLE_TO_WRITE_TO_ROUDI_MQ);
     }
 
+    instanceContainer.clear();
```

Comment:
It is better not to clear the container because it could contain other elements given by the user. 

I would prefer to check if the given ServiceDescription is already exist in the instanceContainer. It is a cxx::vector, so you can easily iterate through it with a range-based loop. This should be done before sendRequestToRouDi(). If the ServiceDescription already exist then just print a Logmessage like ("ServiceDescription already exist") in Debuglevel and return a cxx::success.

Reply:
@dkroenke you are right it is somehow a mess here. What should we do if we get a container which is already full of garbage, then we are unable to return anything. 

I would recommend to change the signature of `findService` to: 
```cpp
cxx::expected<InstanceContainer, Error> PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
```
and return the `InstanceContainer` directly otherwise we have a lot of issues to deal with, like
1. What to do if the provided `InstanceContainer` is full
2. If we found a service should we look in the `InstanceContainer` if its already present.
3. What is with entries which are out of date?

@nihalchari could you please adapt `findService` so that it is returning the `InstanceContainer` directly?

Reply:
yeah @elfenpiff  i could adapt this change, but ill create separate issue and continue. i will link it here

Reply:
In ara::com the instance container is provided as parameter to avoid returning it by value and doing a copy. But yes, we see that it has some drawbacks. As the iceoryx API does not need to follow ara::com here, we can change it. I see only the options a) clear or b) the proposal of @elfenpiff
Are we sure that compilers of today with things like return value optimization are good enough that b) is not way more expensive than a)? The instance container has a max capacity of 50 and an ID string which is a cxx::string<100> I guess. So we are talking about 5 kBytes

Reply:
@budrus I would say we are save here that the compiler will optimize that away. Take a look at the minimalistic example: https://godbolt.org/z/z565x6 . There seems to be only one allocation in assembler (line 32: add rsp, 3888) so that in the end the instance container would actually live in the caller (here main) and is allocated just once and then reused.

Reply:
Yes, container will be reused, but in case of multiple calls container will be overwritten, then it will be callers job to duplicate/store data if necessary… in this case, can it be still achieved by option a) by @budrus  where we clear the container and document in deoxygen, that if user needs initial data from container he needs to store it before passing to findService()

Even if iceoryx API does not follow ara:com, does is it make sense that findService is oneshot call which should only return available instance of requested service at the time of call.
@elfenpiff  @budrus

Reply:
@nihalchari what do you mean with multiple calls?
In the godbolt example the container was not overridden. With every iteration it was recreated again since the end of the loop is the end of the scope and then the scope starts again with the next iteration. But the assembler code showed us that the memory position stays the same and that the compiler optimizes it behind the scene in a way that the memory is reused.

To the oneshot call question: At the moment findService does exactly what you described and if you return the InstanceVector instead of using a reference it will still perform the same thing. I think I do not understand question here either!? :kangaroo:

Reply:
My expectation from a container that was filled/returned by the finService call is that it contains the instances that are now available. Having already some instances in the container and further will be added by the call does not feel right. Therefore I'm not for the option that @dkroenke proposed. Clearing the vector is the workaround to fulfill this contract. But this is also a bit dirty. With the assumption that the compiler will do a good job, the option @elfenpiff proposed seems for me the best solution

Reply:
okay, if clearing container is dirty then i will go with option proposed by @elfenpiff

Reply:
created new issue #387 and disabled failing test

Reply:
changed signature as discussed in this PR, #387 is closed